.github/workflows/esp32.yml

```diff
@@ -0,0 +1,56 @@
+name: CI for ESP32
+
+on:
+  workflow_call:
+
+env:
+  CARGO_INCREMENTAL: 0
+  RUSTFLAGS: "-D warnings --cfg espidf_time32 "
+  RUST_BACKTRACE: short
+
+# by default github use sh as shell
+defaults:
+  run:
+    # will use bash to run each command, bash will source /etc/profile which will give us the environment to build for esp32
+    shell: bash --norc  -leo pipefail {0}
+        
+jobs:
+  all:
+    name: "Build, clippy and test for esp32"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/viamrobotics/micro-rdk-dev-env:amd64
+    steps:
+    - name : Checkout main branch code
+      if: github.event_name != 'pull_request_target'
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 2
+    - name: Check out PR branch code
+      if: github.event_name == 'pull_request_target'
+      uses: actions/checkout@v4
+      with:
+        ref: ${{ github.event.pull_request.head.sha }}
+        fetch-depth: 2
+    # workaround, the environment definition is under the testbot user but github action runs everything as root
+    - name: Move profile
+      run: cp /home/testbot/.bash_profile /etc/profile
+    - name : Set default toolchain to esp
+      run: rustup default esp
+    - name: Cache Dependencies
```

Comment:
we might want to key the cache and only store it when the CI execute from the main branch. Otherwise I am not sure if a cache will be stored for each PR. Since both cache have a combined size of 1.4 GB we can have and handful of PR opens before evicting older cache

Reply:
Isn't the point though to improve PR performance? I don't have a great understanding of how this cache works exactly, and the description on the repo doesn't provide a lot of context.

Reply:
me neither, i guess my main concern is that if we have an handful of PRs open they will evict each other's cache. But i would like to test it this way and see the outcomes

Reply:
Fine with me. I think land this and see how it goes, then we can iterate.